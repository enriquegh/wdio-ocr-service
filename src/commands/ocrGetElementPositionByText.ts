import logger from '@wdio/logger'
import ocrGetTextPositions from '../utils/ocrGetTextPositions'
import { fuzzyFind } from '../utils/fuzzySearch'
import { Rectangles } from '../types/types'
import { SERVICE_NAME } from '../utils/constants'

const log = logger(SERVICE_NAME)

interface OcrGetElementPositionByTextOptions {
  isTesseractAvailable: boolean;
  ocrImagesPath: string;
  reuseOcr: boolean;
  screenSize: {
    width: number;
    height: number;
  };
  text: string;
}

export interface FuzzyElement {
  /**
   * The found item
   */
  item: {
    /**
     * the matched string
     */
    text: string;
    /**
     * The original position
     */
    originalPosition: Rectangles;
    /**
     * The position after DPR check
     * screenshots for iOS are with DPR
     * position on the screen for iOS is smaller
     */
    dprPosition: Rectangles;
  };
  /**
   * Index of the fuzzy logic check
   */
  refIndex: number;
  /**
   * Matched score of the fuzzy logic check
   */
  score: number;
}

interface OcrGetElementPositionByText {
  /**
   * the original search value
   */
  searchValue: string;
  /**
   * the matched string
   */
  matchedString: string;
  /**
   * The original position
   */
  originalPosition: {
    top: number;
    left: number;
    right: number;
    bottom: number;
  },
  /**
   * The position after DPR check
   * screenshots for iOS are with DPR
   * position on the screen for iOS is smaller
   */
  dprPosition: {
    top: number;
    left: number;
    right: number;
    bottom: number;
  },
  /**
   * Matched score of the fuzzy logic check
   */
  score: number;
}

export default async function ocrGetElementPositionByText(
  data: OcrGetElementPositionByTextOptions
): Promise<OcrGetElementPositionByText> {
  const {
    isTesseractAvailable,
    ocrImagesPath,
    reuseOcr,
    screenSize,
    text,
  } = data
  const textPositions = await ocrGetTextPositions({
    isTesseractAvailable,
    ocrImagesPath,
    reuseOcr,
    screenSize,
  })
  const matches = fuzzyFind({
    textArray: textPositions,
    pattern: text,
  })
  let element

  if (matches.length === 0) {
    log.warn(`No matches were found based on the word "${text}"`)

    throw new Error(
      `InvalidSelectorMatch. Strategy 'ocr' has failed to find word '${text}' in the image`
    )
  } else if (matches.length > 1) {
    // @ts-ignore
    matches.sort((a, b) => (a.score > b.score ? 1 : -1))
    const messageOne = `Multiple matches were found based on the word "${text}".`
    // @ts-ignore
    const messageTwo = `The match "${matches[0].item.text}" with score "${(1 - matches[0].score) * 100}%" will be used.`
    log.warn(`${messageOne} ${messageTwo}`)
    element = matches[0] as FuzzyElement
  } else {
    log.info('We found one match')
    element = matches[0] as FuzzyElement
  }

  return {
    searchValue: text,
    matchedString: element.item.text,
    originalPosition: element.item.originalPosition,
    dprPosition: element.item.dprPosition,
    score: Number(((1-element.score)*100).toFixed(2))
  }
}
