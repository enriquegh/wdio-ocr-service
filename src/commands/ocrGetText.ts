import ocrGetData from '../utils/ocrGetData'
import { Rectangles, ScreenSize } from '../types/types'

interface OcrGetTextOptions {
  androidRectangles?:Rectangles;
  iOSRectangles?:Rectangles;
  isTesseractAvailable: boolean;
  ocrImagesPath: string;
  reuseOcr: boolean;
  screenSize: ScreenSize;
}

export default async function ocrGetText(options: OcrGetTextOptions): Promise<string> {
  const { text } = await ocrGetData(options)

  return text.replace(/\n\s*\n/g, '\n')
}
