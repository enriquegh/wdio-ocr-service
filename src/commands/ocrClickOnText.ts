import ocrGetElementPositionByText from './ocrGetElementPositionByText'
import { determineClickPoint } from '../utils'
import { ScreenSize } from '../types/types'

interface OcrClickOnTextOptions {
  isTesseractAvailable: boolean;
  ocrImagesPath: string;
  reuseOcr: boolean;
  screenSize: ScreenSize;
  text: string;
}

export default async function ocrClickOnText(
  options: OcrClickOnTextOptions
): Promise<void> {
  const element = await ocrGetElementPositionByText(options)
  const { x, y } = determineClickPoint({
    rectangles: element.dprPosition,
  })
  await driver.performActions([
    {
      type: 'pointer',
      id: 'finger1',
      parameters: { pointerType: 'touch' },
      actions: [
        {
          type: 'pointerMove',
          duration: 0,
          x,
          y,
        },
        { type: 'pointerDown', button: 0 },
        { type: 'pause', duration: 500 },
        { type: 'pointerUp', button: 0 },
      ],
    },
  ])
}
