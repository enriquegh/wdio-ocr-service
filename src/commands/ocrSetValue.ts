import { Rectangles, ScreenSize } from '../typings/types'
import ocrKeys from '../utils/ocrKeys'
import ocrClickOnText from './ocrClickOnText'

interface OcrSetValueOptions {
  androidRectangles?: Rectangles;
  iOSRectangles?: Rectangles;
  isTesseractAvailable: boolean;
  reuseOcr: boolean;
  ocrImagesPath: string;
  screenSize: ScreenSize;
  text: string;
  value: string;
}

export default async function ocrSetValue(options: OcrSetValueOptions): Promise<void> {
  const {
    androidRectangles,
    iOSRectangles,
    isTesseractAvailable,
    reuseOcr,
    ocrImagesPath,
    screenSize,
    text,
    value,
  } = options

  await ocrClickOnText({
    androidRectangles,
    iOSRectangles,
    isTesseractAvailable,
    ocrImagesPath,
    reuseOcr,
    screenSize,
    text,
  })
  await driver.waitUntil(
    async () => driver.isKeyboardShown(),
    {
      timeout: 15000,
      timeoutMsg: 'Keyboard was not hidden',
    })
  await ocrKeys(value)
  await driver.hideKeyboard()
  await driver.waitUntil(
    async () => !(await driver.isKeyboardShown()),
    {
      timeout: 15000,
      timeoutMsg: 'Keyboard is still shown',
    })
}
