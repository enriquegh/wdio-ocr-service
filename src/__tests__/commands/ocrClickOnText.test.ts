import ocrClickOnText from '../../commands/ocrClickOnText'
import * as OcrGetElementPositionByText from '../../commands/ocrGetElementPositionByText'

describe('ocrClickOnText', () => {
  it('should be able to click on text with the correct arguments', async () => {
    const ocrGetElementPositionByTextSpy = jest.spyOn(OcrGetElementPositionByText, 'default')
      .mockResolvedValue({
        searchValue: '',
        matchedString: '',
        originalPosition: { top: 1, bottom: 2, left: 3, right: 4 },
        dprPosition: { top: 1, bottom: 2, left: 3, right: 4 },
        score: 1,
      })
    const globalAny:any = global
    const options = {
      isTesseractAvailable: true,
      ocrImagesPath: 'ocrImagesPath',
      reuseOcr: true,
      screenSize: { width: 1, height: 2 },
      text: 'text',
    }

    globalAny.driver = {
      performActions: jest.fn(),
    }

    await ocrClickOnText(options)

    expect(ocrGetElementPositionByTextSpy).toHaveBeenCalledTimes(1)
    expect(globalAny.driver.performActions).toHaveBeenCalledWithSnapshot()
  })
})
