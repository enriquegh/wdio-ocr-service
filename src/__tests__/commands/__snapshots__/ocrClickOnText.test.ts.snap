// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ocrClickOnText should be able to click on text with the correct arguments 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "actions": Array [
            Object {
              "duration": 0,
              "type": "pointerMove",
              "x": 3.5,
              "y": 1.5,
            },
            Object {
              "button": 0,
              "type": "pointerDown",
            },
            Object {
              "duration": 500,
              "type": "pause",
            },
            Object {
              "button": 0,
              "type": "pointerUp",
            },
          ],
          "id": "finger1",
          "parameters": Object {
            "pointerType": "touch",
          },
          "type": "pointer",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
